### PR TITLE
Add working detector timeout

### PR DESCRIPTION
**Pull request by cpr-nico**
Add a timeout as a param to avoid turning off the detector if it only is temporarily out of the FOV or out of range.

### CODE REVIEW

File: apriltag_ros/src/continuous_detector.cpp [hunk 1]
```diff
@@ -42,6 +42,11 @@ void ContinuousDetector::onInit ()
   ros::NodeHandle& nh = getNodeHandle();
   ros::NodeHandle& pnh = getPrivateNodeHandle();
 
+  pnh.param<double>("fov_size_scaler", fov_size_scaler_, fov_size_scaler_);
+  pnh.param<double>("min_detection_dist", min_detection_dist_, min_detection_dist_);
+  pnh.param<double>("max_detection_dist", max_detection_dist_, max_detection_dist_);
+  pnh.param<double>("detector_timeout", detector_timeout_, detector_timeout_);
```
Comment (efernandez) [inline apriltag_ros/src/continuous_detector.cpp]: You can simply do:
```suggestion
  pnh.param("fov_size_scaler", fov_size_scaler_, fov_size_scaler_);
  pnh.param("min_detection_dist", min_detection_dist_, min_detection_dist_);
  pnh.param("max_detection_dist", max_detection_dist_, max_detection_dist_);
  pnh.param("detector_timeout", detector_timeout_, detector_timeout_);
```

File: apriltag_ros/src/continuous_detector.cpp [hunk 2]
```diff
@@ -195,16 +202,22 @@ void ContinuousDetector::imageCallback (
...
 
   // Check if not in fov or not in detection range and publish an empty detection message
   if (!isTagInFOV(tag_pose_) || !isTagInDetectionRange(tag_pose_))
   {
-    // ROS_INFO_STREAM_THROTTLE(1.0, "FOV or range check not passed");
-    publishEmptyDetection(image_rect->header);
-    return;
+    if (detection_timeout_)
+    {
+      // ROS_INFO_STREAM_THROTTLE(1.0, "FOV or range check not passed");
```
Comment (efernandez) [inline apriltag_ros/src/continuous_detector.cpp]:
```suggestion
      ROS_DEBUG_STREAM_THROTTLE(1.0, "FOV or range check not passed");
```
:thinking: or maybe it should actully be INFO, or even WARN :thinking:

File: apriltag_ros/src/continuous_detector.cpp [hunk 3]
```diff
@@ -195,16 +202,22 @@ void ContinuousDetector::imageCallback (
...
+      publishEmptyDetection(image_rect->header);
+      return;
+    }
+    
+    detector_validations_failed = true;
   }
 
-  // ROS_INFO_STREAM_THROTTLE(0.1, "Frame " << image_rect->header.frame_id << " is in the FOV, detecting");
+  // ROS_INFO_STREAM_THROTTLE(0.1, "Frame " << image_rect->header.frame_id << " is in the FOV (or within time buffer), detecting");
```
Comment (efernandez) [inline apriltag_ros/src/continuous_detector.cpp]: Remove or make DEBUG if too spammy

File: apriltag_ros/src/continuous_detector.cpp [hunk 4]
```diff
@@ -221,6 +234,14 @@ void ContinuousDetector::imageCallback (
   // Publish detected tags in the image by AprilTag 2
   tag_detections_publisher_.publish(
       tag_detector_->detectTags(cv_image_,camera_info));
+  
+  // only reset the timer if the validations were passed
+  if (!detector_validations_failed)
```
Comment (efernandez) [inline apriltag_ros/src/continuous_detector.cpp]: Do we really need a `ros::Timer` for this?

IMHO it'd be better to have a `ros::Time last_valid_detection_;` attribute that you set to the the `header.stamp` when the detection passes the checks, and then we check `if ((header.stamp - last_valid_detection_).toSec() > detection_timeout_)` where appropriate.

**Review by efernandez — CHANGES_REQUESTED** [review]

File: apriltag_ros/src/continuous_detector.cpp (hunk 3 above)
Comment (cpr-nico) [reply apriltag_ros/src/continuous_detector.cpp]: Yea I removed in the previous MR (Nolan suggested it, the changes just haven't been reflected here yet).

File: apriltag_ros/src/continuous_detector.cpp (hunk 4 above)
Comment (cpr-nico) [reply apriltag_ros/src/continuous_detector.cpp]: Yea I could do it like that too (I agree that it's cleaner), will change

File: apriltag_ros/src/continuous_detector.cpp (hunk 2 above)
Comment (cpr-nico) [reply apriltag_ros/src/continuous_detector.cpp]: no, I'll remove

**Review by efernandez — APPROVED** [review]

**Pull request merged**